### PR TITLE
Don't send entire macro block in consensus

### PR DESCRIPTION
**Pull request by rtsisyk**
- Send only block_hash, base header and new transactions in
   MacroBlockProposal.
- Don't require leader's part to be present in multi-signature of
   sealed block (may happen in case of auto-commit).
- Refactor consensus and remove all generics.
- Rework Consensus <-> NodeService integration.
- Add Transaction:fullhash() method to use during MacroBlock creation.
    
Needed for new macro blocks.

### CODE REVIEW

File: consensus/src/state.rs [hunk 1]
```diff
@@ -267,29 +271,39 @@ impl<Request: Hashable + Clone + Debug + Eq, Proof: Hashable + Clone + Debug>
     }
 
     ///
-    /// Pre-vote the request.
+    /// Returns true if current node should validate a new request.
     ///
-    /// # Arguments
+    pub fn should_prevote(&self) -> Option<(&Hash, &MacroBlockProposal, u32)> {
```
Comment (vldm) [inline consensus/src/state.rs]: Could you correct doc comments, and rename this function. methods started with "should_" should return true / false

Comment (rtsisyk) [reply consensus/src/state.rs]: OK

File: consensus/src/state.rs [hunk 2]
```diff
@@ -96,14 +94,16 @@ pub struct Consensus<Request, Proof> {
     state: ConsensusState,
     /// Current consensus round.
     round: u32,
-    /// Proposed request.
-    request: Option<Request>,
-    /// A proof need to validate request.
-    proof: Option<Proof>,
+    /// The block.
+    block: Option<MacroBlock>,
```
[the remark this answers is not in the document]
Comment (rtsisyk) [reply consensus/src/state.rs]: https://github.com/stegos/stegos/issues/920

[the remark this answers is not in the document]
Comment (rtsisyk) [reply consensus/src/state.rs]: I opened a ticket, but I want to note, that this refactoring requires a lot of efforts. I tried to use `enum-state` pattern for Consensus initially, but failed.

File: node/src/lib.rs [hunk 3]
```diff
@@ -788,14 +788,14 @@ impl NodeService {
...
                             // Check for forks.
-                            let (consensus_block, _proof) = consensus.get_proposal();
-                            let consensus_block_hash = Hash::digest(consensus_block);
-                            if hash != consensus_block_hash {
-                                panic!(
-                                    "Network fork: received_block={:?}, consensus_block={:?}",
-                                    &hash, &consensus_block_hash
-                                );
-                            }
+                            // let (consensus_block, _proof) = consensus.get_proposal();
```
[the remark this answers is not in the document]
Comment (rtsisyk) [reply node/src/lib.rs]: I fixed this check. It works again.

**Pull request merged**